src/tcp_orientation_calibration.cpp

```diff
@@ -0,0 +1,182 @@
+        ROS_INFO("Second point: move the TCP to any location along the X axis of the tool and press enter");
+    else
+        ROS_INFO("Third Point: move TCP to any location along the Y axis of the tool and press enter.");
+    std::getline(std::cin, line); // Blocks program until enter is pressed
+    tf::StampedTransform transform;
+    try
+    {
+      listener.lookupTransform(base_frame, tool0_frame, ros::Time(0), transform);
+      Eigen::Affine3d eigen_pose;
+      tf::poseTFToEigen(transform, eigen_pose); // convert transform to Affine3d
+      A0n[count] = eigen_pose;
+      ROS_INFO_STREAM("Pose " << i << ": captured transform:\n" << eigen_pose.matrix());
+    }
+    catch (const tf::TransformException& ex)
+    {
+      ROS_ERROR("%s",ex.what());
+      ros::Duration(1.0).sleep();
+      continue;
+    }
+  }
+  // here we have the three  transformation A0n_1st, A0n_2nd, A0n_3rd
+    Vector4d r0ns_n = {tcp_transl[0], tcp_transl[1], tcp_transl[2], 1}; //tool center popint vector from pkg_1
+    Vector4d rd_0 = A0n[0]*r0ns_n;
+    Vector4d r0ns_0 = rd_0;
+    //  point 2: calculate delat1 and then c11, c21, c31
+    Vector4d V1;//{ vx, vy, vz, 1};
+    V1 = A0n[1].inverse() * A0n[0] * r0ns_n; // r0ns_0 has been modified to r0ns_n
+    double term1 = ( V1(0) - r0ns_n(0) );
+    double term2 = ( V1(1) - r0ns_n(1) );
+    double term3 = ( V1(2) - r0ns_n(2) );
+    double delta1 = -1* sqrt( term1*term1 + term2*term2 + term3*term3);
+    double c11 = term1 / delta1;
+    double c21 = term2 / delta1;
+    double c31 = term3 / delta1;
+    ROS_INFO_STREAM(" delat1: "<< delta1);
+    std::cout<<"c11, c21, c31: "<< c11 << " "<< c21 << " " <<c31<<std::endl;
+    //  point 3: calculate delta2 and then c12, c22, c32
+     Vector4d V2; //={ vx, vy, vz, 1};
+     V2 = A0n[2].inverse() * A0n[0] * r0ns_n; // r0ns_0 has been modified to r0ns_n
+     term1 = ( V2(0) - r0ns_n(0) );
+     term2 = ( V2(1) - r0ns_n(1) );
+     term3 = ( V2(2) - r0ns_n(2) );
+     double delta2 = 1* sqrt(term1*term1 + term2*term2 + term3*term3);
+     double c12 = term1 / delta2;
+     double c22 = term2 / delta2;
+     double c32 = term3 / delta2;
+     ROS_INFO_STREAM(" delat2: "<< delta2);
+     std::cout<<"c21, c22, c23: "<< c12 << " "<< c22 << " " << c32 <<std::endl;
+     //calculate c31, c32, c33;  c13=k1.c33  && c23=k2.c33
+     double k1= (c21*c32 - c22*c31)/(c11*c22 - c21*c12);
+     double k2= (c12*c31 - c11*c31)/(c11*c22 - c21*c12);
+     double c33 = sqrt( 1 /(1+k1*k1+k2*k2) );
+     double c13 = k1* c33;
+     double c23 = k2* c33;
+     std::cout<<"c31, c32, c32: "<< c31 << " "<< c32 << " " <<c33<<std::endl;
+     // store rotation matrix and translation
+     Matrix3d Rot;
+              Rot << c11, c12, c13,
+                     c21, c22, c23,
+                     c31, c32, c33;
+     Vector3d transl= {r0ns_n(0), r0ns_n(1), r0ns_n(2)};
+     std::cout<<"Rot_matrix:\n "<< Rot <<std::endl;
+     std::cout<<"translation:\n "<< transl <<std::endl;
```

Comment:
Can you use `ROS_INFO_STREAM` or `ROS_DEBUG_STREAM` messages here?

Reply:
done!